qiskit/circuit/quantumcircuit.py

```diff
@@ -3927,6 +3928,196 @@ def pauli(
+                )
+        elif isinstance(condition[0], Clbit):
+            if condition[0] not in self.clbits:
+                raise CircuitError(
+                    "The condition provided to QuantumCircuit.if_ was not "
+                    "found on the enclosing circuit. "
+                    f"condition: {condition}, QuantumCircuit.clbits: {self.clbits}."
+                )
+
+        return self.append(IfElseOp(condition, true_body, None, label), qubits, clbits)
+
+    def if_else(
+        self,
+        condition: Union[
+            Tuple[ClassicalRegister, int],
+            Tuple[Clbit, int],
+            Tuple[Clbit, bool],
+        ],
+        true_body: "QuantumCircuit",
+        false_body: "QuantumCircuit",
+        qubits: Sequence[QubitSpecifier],
+        clbits: Sequence[ClbitSpecifier],
+        label: Optional[str] = None,
+    ) -> InstructionSet:
+        """Apply :class:`~qiskit.circuit.controlflow.IfElseOp`.
+
+        Args:
+            condition: A condition to be evaluated at circuit runtime which,
+                if true, will trigger the evaluation of ``true_body``. Can be
+                specified as either a tuple of a ``ClassicalRegister`` to be
+                tested for equality with a given ``int``, or as a tuple of a
+                ``Clbit`` to be compared to either a ``bool`` or an ``int``.
+            true_body: The circuit body to be run if ``condition`` is true.
+            false_body: The circuit to be run if ``condition`` is false.
+            qubits: The circuit qubits over which the if/else should be run.
+            clbits: The circuit clbits over which the if/else should be run.
+            label: The string label of the instruction in the circuit.
+
+        Raises:
+            CircuitError: If the provided condition references Clbits outside the
+                enclosing circuit.
+
+        Returns:
+            A handle to the instruction created.
+        """
+        # pylint: disable=cyclic-import
+        from qiskit.circuit.controlflow.if_else import IfElseOp
+
+        if isinstance(condition[0], ClassicalRegister):
+            if condition[0] not in self.cregs:
+                raise CircuitError(
+                    "The condition provided to QuantumCircuit.if_else was not "
+                    "found on the enclosing circuit. "
+                    f"condition: {condition}, QuantumCircuit.cregs: {self.cregs}."
+                )
+        elif isinstance(condition[0], Clbit):
+            if condition[0] not in self.clbits:
+                raise CircuitError(
+                    "The condition provided to QuantumCircuit.if_else was not "
+                    "found on the enclosing circuit. "
+                    f"condition: {condition}, QuantumCircuit.clbits: {self.clbits}."
+                )
+
+        return self.append(IfElseOp(condition, true_body, false_body, label), qubits, clbits)
```

Comment:
Do we require that the condition bit must be in the `clbits` argument as well?  I think we need it, in order to have the blocking work correctly?  Either way, `self.clbits` is a `List`, so we should make sure we do the check in a set-like structure - we'll have poor scaling in a wide circuit that's deep in `if_else` (and `if_` and `while_loop`) statements.

Reply:
Not strictly, here or elsewhere (though I think this should change). The benefit of unifying `WhileLoopOp` and `ForLoopOp` under `Instruction.condition` is that the blocking is already handled (via the existing machinery to support `c_if`.)

Good point on set lookup. Added in f681c37c .